source/common/network/listener_impl.cc

```diff
@@ -37,11 +41,11 @@ void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr*
     // Hands off redirected connections (from iptables) to the listener associated with the
     // original destination address. If there is no listener associated with the original
     // destination address, the connection is handled by the listener that receives it.
-    // Note: A listener that has the use_original_dst flag set to true can still receive.
+    // Note: A listener that has the use_original_dst flag set to true can still receive
     // connections that are NOT redirected using iptables. If a connection was not redirected,
-    // the address returned by getOriginalDst() match the listener address. In this case the
-    // listener handles the connection directly and does not hand it off.
-    if (listener->socket_.localAddress() != final_local_address) {
+    // the address returned by getOriginalDst() matches the local address of the new socket.
+    // In this case the listener handles the connection directly and does not hand it off.
+    if (!(*final_local_address == *local_address)) {
```

Comment:
Right, considered that as well but then could not decide which way is the best to do it. Typically operator!= would be defined in terms of the operator== using the '!(a == b)' construct I used here. All the member of these base classes are pure virtual, but I thought calling another trivial pure virtual member from the first one is kind of excessive, unless the compiler can optimize that case somehow. Could also define the operator!= as non-virtual, since the operator== is already virtual, but I haven't seen a precedence for that in Envoy source yet. Would a non-virtual member operator!= that always calls the virtual operator== be acceptable?